### PR TITLE
Input inhibit

### PR DESCRIPTION
**Pull request by SirCmpwn**
Ref https://github.com/swaywm/wlr-protocols/pull/9

Test plan:

1. Open various combinations of shell surfaces and layer surfaces
2. `./examples/input-inhibitor`
3. Provide input and ensure all of it goes to input-inhibitor and none of it anywhere else

Ideally test touch pointer, keyboard, touch, and tablet input.

### CODE REVIEW

**Comment by SirCmpwn** [comment]
I recommend reviewing this one commit at a time

File: include/wlr/types/wlr_input_inhibitor.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#ifndef _WLR_TYPES_INPUT_INHIBITOR_H
+#define _WLR_TYPES_INPUT_INHIBITOR_H
```
Comment (emersion) [inline include/wlr/types/wlr_input_inhibitor.h]: We decided to use a convention without a leading underscore.

Comment (SirCmpwn) [reply include/wlr/types/wlr_input_inhibitor.h]: Wait when did we decide that

File: include/wlr/types/wlr_input_inhibitor.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+struct wlr_input_inhibit_manager {
+	struct wl_global *wl_global;
+	struct wl_client *active_client;
+	struct wl_resource *active_inhibitor;
+
+	struct wl_listener display_destroy;
+
+	struct {
+		struct wl_signal activate;   // struct wlr_input_inhibit_manager *
+		struct wl_signal deactivate; // struct wlr_input_inhibit_manager *
```
Comment (emersion) [inline include/wlr/types/wlr_input_inhibitor.h]: +1 to adding signal data types

Comment (SirCmpwn) [reply include/wlr/types/wlr_input_inhibitor.h]: Idea came from #776

File: include/wlr/types/wlr_input_inhibitor.h (hunk 1 above)
Comment (emersion) [reply include/wlr/types/wlr_input_inhibitor.h]: https://github.com/swaywm/wlroots/pull/157

File: types/wlr_input_inhibitor.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	struct wl_resource *wl_resource = wl_resource_create(client,
+			&zwlr_input_inhibitor_v1_interface,
+			wl_resource_get_version(resource), id);
+	if (!wl_resource) {
+		wl_client_post_no_memory(client);
+	}
+	wl_resource_set_implementation(wl_resource, &input_inhibitor_implementation,
+			manager, NULL);
```
Comment (emersion) [inline types/wlr_input_inhibitor.c]: What happens if the client crashes or disconnects without sending a `destroy` request?

I'd move the `input_inhibitor_destroy` logic into the `wl_resource` destroy signal handler.

Comment (SirCmpwn) [reply types/wlr_input_inhibitor.c]: >What happens if the client crashes or disconnects without sending a destroy request?

https://github.com/swaywm/wlroots/pull/807/files/0e613bed5cf0383450586de9b9219fbf639e8db7#diff-73e9f66aeebd8121768ea744a5bff64fR61

File: types/wlr_input_inhibitor.c [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	wl_signal_init(&manager->events.deactivate);
+
+	manager->display_destroy.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &manager->display_destroy);
+
+	manager->wl_global = wl_global_create(display,
+			&zwlr_input_inhibit_manager_v1_interface,
+			1, manager, inhibit_manager_bind);
+	if (manager->wl_global == NULL){
+		wl_list_remove(&manager->display_destroy.link);
```
Comment (emersion) [inline types/wlr_input_inhibitor.c]: What about setting up the `display_destroy` listener after this `if`?

File: rootston/cursor.c [hunk 5]
```diff
@@ -328,8 +332,12 @@ void roots_cursor_handle_touch_down(struct roots_cursor *cursor,
 
 	uint32_t serial = 0;
 	if (surface) {
-		serial = wlr_seat_touch_notify_down(cursor->seat->seat, surface,
-			event->time_msec, event->touch_id, sx, sy);
+		if (cursor->seat->exclusive_client &&
+				wl_resource_get_client(surface->resource) !=
+				cursor->seat->exclusive_client) {
```
Comment (emersion) [inline rootston/cursor.c]: Seems like this condition is inverted

File: rootston/cursor.c [hunk 6]
```diff
@@ -379,10 +387,14 @@ void roots_cursor_handle_touch_motion(struct roots_cursor *cursor,
 			desktop, lx, ly, &sx, &sy, NULL);
 
 	if (surface) {
-		wlr_seat_touch_point_focus(cursor->seat->seat, surface,
-			event->time_msec, event->touch_id, sx, sy);
-		wlr_seat_touch_notify_motion(cursor->seat->seat, event->time_msec,
-			event->touch_id, sx, sy);
+		if (cursor->seat->exclusive_client &&
+				wl_resource_get_client(surface->resource) !=
+				cursor->seat->exclusive_client) {
```
Comment (emersion) [inline rootston/cursor.c]: Ditto

File: rootston/seat.c [hunk 7]
```diff
@@ -717,6 +719,11 @@ struct roots_seat_view *roots_seat_view_from_view(
 }
 
 void roots_seat_set_focus(struct roots_seat *seat, struct roots_view *view) {
+	if (view && seat->exclusive_client && wl_resource_get_client(
+				view->wlr_surface->resource) != seat->exclusive_client) {
```
Comment (emersion) [inline rootston/seat.c]: Maybe adding a `roots_seat_accept_input(struct wl_resource *resource)` could be better

File: wlroots.syms [hunk 8]
```diff
@@ -3,6 +3,7 @@ WLROOTS_0_0_0 {
         wlr_*;
         _wlr_log;
         _wlr_vlog;
+		usec_to_msec;
```
Comment (emersion) [inline wlroots.syms]: Do we really want to export this one?

File: rootston/seat.c [hunk 9]
```diff
@@ -716,7 +718,17 @@ struct roots_seat_view *roots_seat_view_from_view(
 	return seat_view;
 }
 
+bool roots_seat_allow_input(struct roots_seat *seat,
+		struct wl_resource *resource) {
+	return !seat->exclusive_client ||
+			wl_resource_get_client(resource) != seat->exclusive_client;
```
Comment (emersion) [inline rootston/seat.c]: `==`

**Pull request merged**